### PR TITLE
Act 165 create api to follow and unfollow event

### PR DESCRIPTION
**Pull request by TranSiTien**

### CODE REVIEW

File: src/main/resources/db/migration/V242310_01__create_event_follow_table.sql [hunk 1]
```diff
@@ -0,0 +1,19 @@
+BEGIN;
+
+CREATE TABLE event_follow
+(
+    id            SERIAL PRIMARY KEY,
+    follow_at TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP,
```
Comment (Sang-Tan) [inline src/main/resources/db/migration/V242310_01__create_event_follow_table.sql]:
```suggestion
    followed_at TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP,
```

File: src/main/resources/db/migration/V242310_01__create_event_follow_table.sql [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    student_id    INTEGER,
+    event_id      INTEGER
```
Comment (Sang-Tan) [inline src/main/resources/db/migration/V242310_01__create_event_follow_table.sql]:
```suggestion
    student_id    INTEGER NOT NULL,
    event_id      INTEGER NOT NULL
```

File: src/main/java/com/dutact/web/core/entities/event/EventStatus.java [hunk 3]
```diff
@@ -19,12 +19,12 @@ public abstract class EventStatus {
...
     @Setter
     public static class Approved extends EventStatus {
-        public static final String TYPE_NAME = "approved";
-        public String moderatedAt;
-
         public Approved() {
             this.moderatedAt = LocalDateTime.now().toString();
         }
+
+        public static final String TYPE_NAME = "approved";
```
Comment (Sang-Tan) [inline src/main/java/com/dutact/web/core/entities/event/EventStatus.java]: Why move this

File: src/main/java/com/dutact/web/core/entities/event/EventStatus.java [hunk 4]
```diff
@@ -34,15 +34,14 @@ public static class Pending extends EventStatus {
...
     public static class Rejected extends EventStatus {
         public static final String TYPE_NAME = "rejected";
-        private String reason;
-        private String moderatedAt;
 
         public Rejected(String reason) {
             this.reason = reason;
             this.moderatedAt = LocalDateTime.now().toString();
         }
+        private String reason;
```
Comment (Sang-Tan) [inline src/main/java/com/dutact/web/core/entities/event/EventStatus.java]: Why move this

File: src/main/java/com/dutact/web/core/entities/event/EventStatus.java [hunk 5]
```diff
@@ -34,15 +34,14 @@ public static class Pending extends EventStatus {
 
     @Getter
     @Setter
-    @NoArgsConstructor
```
Comment (Sang-Tan) [inline src/main/java/com/dutact/web/core/entities/event/EventStatus.java]: Why move this

File: src/main/java/com/dutact/web/features/event/student/controllers/StudentEventController.java [hunk 6]
```diff
@@ -87,4 +89,47 @@ public ResponseEntity<?> unregister(@PathVariable Integer id)
...
+        Integer requestStudentId = studentAccountService
+                .getStudentId(SecurityContextUtils.getUsername())
+                .orElseThrow(() ->
+                        new RuntimeException("The account is not associated with any student profile"));
+
+        try {
+            return ResponseEntity.ok(eventService.follow(id, requestStudentId));
+        } catch (AlreadyFollowedException e) {
+            return ResponseEntity.status(HttpStatus.CONFLICT)
+                    .body(Map.of("error", "Student already follow this event"));
```
Comment (Sang-Tan) [inline src/main/java/com/dutact/web/features/event/student/controllers/StudentEventController.java]: Use the errormessage object

**Review by Sang-Tan — CHANGES_REQUESTED** [review]

File: src/main/resources/db/migration/V242310_01__create_event_follow_table.sql (hunk 2 above)
Comment (TranSiTien) [reply src/main/resources/db/migration/V242310_01__create_event_follow_table.sql]: oke

File: src/main/java/com/dutact/web/core/entities/EventFollow.java [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+@Getter
+@Setter
+public class EventFollow {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    @Column(name = "id")
+    private Integer id;
+
+    @Column(name = "followed_at")
+    private LocalDateTime followAt = LocalDateTime.now();
```
Comment (Sang-Tan) [inline src/main/java/com/dutact/web/core/entities/EventFollow.java:21]:
```suggestion
    private LocalDateTime followedAt = LocalDateTime.now();
```

File: src/main/java/com/dutact/web/core/entities/event/EventStatus.java [hunk 8]
```diff
@@ -39,10 +38,10 @@ public static class Rejected extends EventStatus {
         public static final String TYPE_NAME = "rejected";
         private String reason;
         private String moderatedAt;
-
         public Rejected(String reason) {
             this.reason = reason;
             this.moderatedAt = LocalDateTime.now().toString();
         }
+
```
Comment (Sang-Tan) [inline src/main/java/com/dutact/web/core/entities/event/EventStatus.java]:
```suggestion
```

File: src/main/resources/db/migration/V242310_01__create_event_follow_table.sql [hunk 9]
```diff
@@ -0,0 +1,19 @@
+BEGIN;
+
+CREATE TABLE event_follow
+(
+    id            SERIAL PRIMARY KEY,
+    followed_at TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP,
```
Comment (Sang-Tan) [inline src/main/resources/db/migration/V242310_01__create_event_follow_table.sql:6]: format code

**Pull request merged**